# PART I   < Python程式設計 >
## 2. 資料型別與基本運算
- [2.1 數值運算](#數值運算)
- [2.2 字串運算](#字串運算)


<a id='數值運算'></a>
### 2.1  數值運算
- int
- float
- Decimal
- Fraction
- complex number



In [1]:
2 + 28         # 30

30

In [2]:
150 - 5*26     # 20

20

#### Division (/) always returns a floating point number.

In [3]:
(150 - 5*26) / 5    # 4.0

4.0

In [4]:
80 / 5 

16.0

#### floor division //

In [5]:
79 // 3

26

#### modulus operator %

In [6]:
79 % 3  

1

####  ** operator to calculate powers 

In [7]:
1.41421 ** 2

1.9999899240999999

In [8]:
10**3

1000

#### Assignment Operator =

In [9]:
interest_rate = 0.06
interest_rate

0.06

In [10]:
capital = 1000000
1 + interest_rate

1.06

#### The last printed expression is assigned to the variable _.

In [11]:
_ ** 32

6.4533866818853385

In [12]:
capital * _

6453386.681885338

### `Decimal` objects
- https://docs.python.org/3/library/decimal.html#decimal.Decimal

#### [NOTE]: There are some small differences between arithmetic on Decimal objects and arithmetic on integers and floats. 

In [13]:
(-11) % 4    ##  the remainder of (-11) % 4

1

In [14]:
(-11) // 4   ##  the quotiant of (-11) % 4

-3

- When the remainder operator % is applied to Decimal objects, the sign of the result is the sign of the dividend rather than the sign of the divisor:

In [15]:
from decimal import Decimal
Decimal(-11) % Decimal(4)

Decimal('-3')

- Similarly for // :

In [16]:
Decimal(-11) // Decimal(4)

Decimal('-2')

### `Fraction` objects
- https://docs.python.org/3/library/fractions.html#fractions.Fraction

In [17]:
from fractions import Fraction
Fraction(30, -4)

Fraction(-15, 2)

In [18]:
Fraction(2018)

Fraction(2018, 1)

In [19]:
Fraction()

Fraction(0, 1)

In [20]:
Fraction('4/23')

Fraction(4, 23)

In [21]:
Fraction('-4/23')

Fraction(-4, 23)

#### a finite floating number => a rational number = int / int

In [22]:
Fraction('1.41421')  

Fraction(141421, 100000)

In [23]:
Fraction('2.71828')

Fraction(67957, 25000)

In [24]:
Fraction('-.125')

Fraction(-1, 8)

In [25]:
Fraction('1.7e-3')

Fraction(17, 10000)

In [26]:
Fraction(3.33)

Fraction(1874623344892969, 562949953421312)

In [27]:
from decimal import Decimal
Fraction(Decimal('3.33'))    ##  Why?

Fraction(333, 100)

#### Why Fraction(333, 100) ? 
#### Ans : The Fraction class inherits from the abstract base class `numbers.Rational`, and implements all of the methods and operations from that class.

<a id='字串運算'></a>
### 2.2  字串運算

In [28]:
'Hello Python!'  # single quotes

'Hello Python!'

In [29]:
'isn\'t'  # use \' to escape the single quote...

"isn't"

In [30]:
"isn't"  # ...or use double quotes instead

"isn't"

In [31]:
'"Could you give me a ride, Alice?" Bob said.'

'"Could you give me a ride, Alice?" Bob said.'

In [32]:
"\"Sure. No problem.\" Alice answered."

'"Sure. No problem." Alice answered.'

In [33]:
'"No, it isn\'t," he said.'

'"No, it isn\'t," he said.'

#### Printing the string with `print()` function.

In [34]:
print('"No, it isn\'t," he said.')

"No, it isn't," he said.


In [35]:
string = 'Hello, Alice.\nHi, Bob. How are you?'  # \n means newline
string  

'Hello, Alice.\nHi, Bob. How are you?'

In [36]:
print(string)  # with print(), \n produces a new line

Hello, Alice.
Hi, Bob. How are you?


#### Use raw strings by adding an `r` before the first quote to avoid the escape character `\` :

In [37]:
print('C:\Python\notebook')  

C:\Python
otebook


In [38]:
print(r'C:\Python\notebook') 

C:\Python\notebook


#### Using triple-quotes: `"""..."""` or `'''...'''` to span string literals into multiple lines.

In [39]:
print("""\
[Agenda]
     9:00 ~  9:30               Reception
     9:30 ~ 10:20               Keynote Speech
""")

[Agenda]
     9:00 ~  9:30               Reception
     9:30 ~ 10:20               Keynote Speech



#### Strings' concatenation (+)  and repeatition (*):

In [40]:
'repea' + 'ti'*2 + 'on' 

'repeatition'

#### Two or more string literals next to each other are automatically concatenated.

In [41]:
'He' 'llo' ' Python' 

'Hello Python'

In [42]:
string = ("Fly Robin Fly" " Up Up To The Sky")
string

'Fly Robin Fly Up Up To The Sky'

In [43]:
string "Python"  # Error : can't concatenate a variable and a string literal

SyntaxError: invalid syntax (<ipython-input-43-32f11543abd3>, line 1)

In [44]:
'repea' ('ti'*2) 'on'     ##  'repea' + 'ti'*2 + 'on' 

SyntaxError: invalid syntax (<ipython-input-44-0d17f6c1facc>, line 1)

In [45]:
text = 'Hello Python!'
text[0]

'H'

In [46]:
text[3]

'l'

In [47]:
text[12]

'!'

In [48]:
text[-1]

'!'

In [49]:
text[-7]

'P'

In [50]:
text[-8]

' '

In [51]:
text[-0]

'H'

In [52]:
text[20]

IndexError: string index out of range

In [53]:
text[3:20]

'lo Python!'

In [54]:
text[20:]

''

#### Python strings are `immutable`; hence, assigning to an indexed position in the string results in an error:

In [55]:
text[5] = 's'  ##  Error: 'str' object does not support item assignment !!

TypeError: 'str' object does not support item assignment

####  Creating a new string...

In [56]:
text[:6] + 'world!'

'Hello world!'

In [57]:
text[6:10] + "agoras"

'Pythagoras'

#### 字串長度 - len()

In [58]:
print(text)
print(text[12])
len(text)

Hello Python!
!


13

### `str` function

In [59]:
str(1000)   ##  Transform an integer into a string

'1000'

In [60]:
str(3.1415926) ##  Transform a floating point into a string

'3.1415926'